Mixer - Add separators, fine-tune spacing

cbjeukendrup opened the pull request:
Resolves: #9149
Resolves: #9634
Resolves: #9246

- Added separator lines between channel strips in Mixer
- Fine-tuned the spacing
- You can now also double-click the handle of the volume slider to reset the volume
- To change the value of the pan knob, you now need to drag up and down instead of drawing circles with the mouse

Asmatzaile commented:
Hi,
I don't know about Ableton but some DAWs, such as Reaper, allow left-right scrolling in addition to down-up scrolling to set the value of the pan knob* (I pan scrolling in the former way because it feels more natural to me).
I can't get to see the mixer in the Windows artifacts generated by this PR, but by looking at the code it seems that it only accepts the mouse y value as input. Could you make it so the mouse's displacement in the x axis is also taken into in account when setting the knob's value?
Many thanks

*and knobs in general

jessjwilliamson commented:
> Hi, I don't know about Ableton but some DAWs, such as Reaper, allow left-right scrolling in addition to down-up scrolling to set the value of the pan knob* (I pan scrolling in the former way because it feels more natural to me). I can't get to see the mixer in the Windows artifacts generated by this PR, but by looking at the code it seems that it only accepts the mouse y value as input. Could you make it so the mouse's displacement in the x axis is also taken into in account when setting the knob's value? Many thanks
> 
> *and knobs in general

Hi! It may not have been clear enough from my end, but the Y value was suppose to be an ADDITION to the already implemented value, so both types of panning was the desired way. Can see on my [Ableton demonstration](https://github.com/musescore/MuseScore/issues/9246) that it does both - going around the edge AS WELL as the up and down motion :)

jessjwilliamson commented:
Just a couple notes:
- I mentioned on a reply above about the panning, can it be possible to still keep the original function and have the up and down as well? See [Ableton for demonstration](https://github.com/musescore/MuseScore/issues/9246) 

> Hi, I don't know about Ableton but some DAWs, such as Reaper, allow left-right scrolling in addition to down-up scrolling to set the value of the pan knob* (I pan scrolling in the former way because it feels more natural to me). I can't get to see the mixer in the Windows artifacts generated by this PR, but by looking at the code it seems that it only accepts the mouse y value as input. Could you make it so the mouse's displacement in the x axis is also taken into in account when setting the knob's value? Many thanks
> 
> *and knobs in general

Hi, Sorry I made a comment then deleted it because I was mistaken - I thought Ableton does both Y  value and also around, but it looks like it's just the Y value. Just checked Ableton, Reaper and Logic, they both have this Y value, they don't really have a way to go 'around' and would prefer to keep it that way as it is the standard now.

jessjwilliamson commented:
@cbjeukendrup 
Hi! Thank you so much for implementing these! just a couple of things:
1. Can the panning also go back to 0 value when double-clicking on the pan knob as well please! Just like the volume fader, just panning as well :D 
2. Unfortunately, the strokes do not go all the way to the top when resizing the whole Mixer window vertically. Would be great that the strokes can go all the way up no matter how tall the Mixer window is
![image](https://user-images.githubusercontent.com/71656189/144228840-cb3c5ae2-cefd-4c19-9535-3e04228ef57b.png)
3. When adding Audio FX, there are missing stroke gaps on channels where there are no FX
![Screenshot 2021-12-01 115201](https://user-images.githubusercontent.com/71656189/144229808-72234be8-335c-48e8-b57e-3a7bdb7c7282.png)
That is all!

DmitryArefiev commented:
Hi @cbjeukendrup , I think volume fader shouldn't behave like this when dragging:

https://user-images.githubusercontent.com/90187801/144230672-b8c9b40c-288c-4bb4-89e0-ce2da044f47a.mp4

Asmatzaile commented:
> Hi, Sorry I made a comment then deleted it because I was mistaken - I thought Ableton does both Y value and also around, but it looks like it's just the Y value. Just checked Ableton, Reaper and Logic, they both have this Y value, they don't really have a way to go 'around' and would prefer to keep it that way as it is the standard now.

So yes, among the DAWs I checked and had knobs (Ableton, Logic, Reaper, LMMS, FL Studio), all accept the Y value and only Reaper accepts the X value also. Worth noting that other DAWs that don't have knobs (for example, Cubase, Ardour and Audacity, or Dorico's mixer) use X displacement for panning.
So, even though adding horizontal displacement would feel more natural for me and maybe some other users (after all, drag left to pan left, and drag right to pan right), and it wouldn't interfere with the workflow of people accustomed to dragging vertically, I understand that to resemble the knob's behavior in most other applications you'd want to ignore the movement in the X axis.

cbjeukendrup commented:
@jessjwilliamson @DmitryArefiev I fixed all these issues! And two more:

- Improved appearance of FX slots
- All channels will show an equal number of FX slots, so when some channels have less filled slots than others, those channels will have multiple empty slots. The maximum number of 4 slots still applies.

Next round of testing can begin!

jessjwilliamson commented:
> @jessjwilliamson @DmitryArefiev I fixed all these issues! And two more:
> 
> * Improved appearance of FX slots
> * All channels will show an equal number of FX slots, so when some channels have less filled slots than others, those channels will have multiple empty slots. The maximum number of 4 slots still applies.
> 
> Next round of testing can begin!

Check your discord messages about the empty slots please! :)

jessjwilliamson commented:
@cbjeukendrup Here are the next couple of UI steps to look into now :)
1. Sounds are 26 pixels high and Audio FX are 24 (if that becomes a height problem in the future then all should be 24)
2. Take a look into the hover states of each button of the Sounds and FX slots - bypass button, open VST and dropdown. At the moment it's acting that a whole button is 1 hover state, but hit states are so far correct. If you go the the 'MS4 Main Interface Panels' project I've highlighted with a green box to show where to find the state rules for you to look at :) 

Thank you!

cbjeukendrup commented:
@jessjwilliamson Fixed!

@vpereverzev I heard that you had deliberately not implemented "showing multiple empty fx slots when other channels have more filled slots" because that would cause problems. Could you tell me what problems that would cause?

I implemented it in this PR, but my implementation is probably naive and inefficient, so probably it's better if I remove it from this PR and either me or someone else creates a more decent solution. 

Of course it is undesired that there is an effect in the second slot but not one in the first slot. That situation could occur with this PR. But that situation can also occur now already, namely when the user adds two effects and then removes the first one but keeps the second one (although I have not been able to produce that situation, because the app crashes almost always when removing an effect).

So, what do you think I should do with the solution in this PR?

vpereverzev commented:
> @jessjwilliamson Fixed!
> 
> @vpereverzev I heard that you had deliberately not implemented "showing multiple empty fx slots when other channels have more filled slots" because that would cause problems. Could you tell me what problems that would cause?
> 
> I implemented it in this PR, but my implementation is probably naive and inefficient, so probably it's better if I remove it from this PR and either me or someone else creates a more decent solution.
> 
> Of course it is undesired that there is an effect in the second slot but not one in the first slot. That situation could occur with this PR. But that situation can also occur now already, namely when the user adds two effects and then removes the first one but keeps the second one (although I have not been able to produce that situation, because the app crashes almost always when removing an effect).
> 
> So, what do you think I should do with the solution in this PR?

Ye, it was an intentional thing. In short, it's all about playing around the performance and efficiency. And even the current state is not as performant, as I want it to be.

So I'd encourage you to exclude any changes to the models and limit yourself with QML changes

cbjeukendrup commented:
> Ye, it was an intentional thing. In short, it's all about playing around the performance and efficiency. And even the current state is not as performant, as I want it to be.
> 
> So I'd encourage you to exclude any changes to the models and limit yourself with QML changes

Thanks, I removed that commit. I only kept the changes to the model that were necessary for the QML things (and renamed `ItemRole` to `ChannelItemRole`, since that's slightly more descriptive and was necessary to avoid naming conflicts with QML `Loader`'s `item` property), but those are trivial.

So I think this PR is ready for review then! (It's probably the easiest to look at the commits in isolation.)

vpereverzev commented:
> > Ye, it was an intentional thing. In short, it's all about playing around the performance and efficiency. And even the current state is not as performant, as I want it to be.
> > So I'd encourage you to exclude any changes to the models and limit yourself with QML changes
> 
> Thanks, I removed that commit. I only kept the changes to the model that were necessary for the QML things (and renamed `ItemRole` to `ChannelItemRole`, since that's slightly more descriptive and was necessary to avoid naming conflicts with QML `Loader`'s `item` property), but those are trivial.
> 
> So I think this PR is ready for review then! (It's probably the easiest to look at the commits in isolation.)

OK, I'll take a look tomorrow

vpereverzev commented on src/playback/qml/MuseScore/Playback/internal/AudioResourceControl.qml:
```diff
@@ -241,17 +235,17 @@ Item {
             Layout.alignment: Qt.AlignRight
 
-            active: (root.showAdditionalButtons || root.resourcePickingActive) && root.supportsMenu
+            visible: root.showAdditionalButtons && root.supportsMenu
```
not quite sure about this change, seems redundant

cbjeukendrup replied:
Do you mean the change from `active` to `visible`? Mainly because I think it is semantically more correct, but also because Layout automatically excludes items that are invisible when it calculates the position of other items. So we don't need to set `Layout.preferredWidth` to zero manually (using property binding). That has the advantage that we can bind other properties to `Layout.preferredWidth`. But you could also argue that it is no problem when `Layout.preferredWidth` is set to zero when the loader is inactive, because then the items that would bind properties to it are not rendered anyway... 

Or if you mean that I removed `|| root.resourcePickingActive`, that is because I moved that inside the `root.showAdditionalButtons` property, because I found that we're using `root.showAdditionalButtons || root.resourcePickingActive` again and again.

vpereverzev replied:
Ha, haven't seen this "root.showAdditionalButtons" before, who made this change?

Looks fine to me anyway

vpereverzev commented on src/playback/view/internal/abstractaudioresourceitem.cpp:
```diff
@@ -31,6 +31,11 @@ bool AbstractAudioResourceItem::isBlank() const
 
+bool AbstractAudioResourceItem::isActive() const
+{
+    return true;
```
it's better to make it 'false' by default

vpereverzev commented on src/playback/view/mixerpanelmodel.cpp:
```diff
@@ -34,6 +34,10 @@ MixerPanelModel::MixerPanelModel(QObject* parent)
         load(QVariant::fromValue(m_itemsNavigationSection));
     });
+
+    connect(this, &QAbstractItemModel::modelReset, this, &MixerPanelModel::rowCountChanged);
```
I think you could just emit countChanged at "add/removeItem" methods

cbjeukendrup commented on src/playback/qml/MuseScore/Playback/MixerPanel.qml:
```diff
@@ -75,8 +78,8 @@ Rectangle {
 
-        contentWidth: contentColumn.width
-        contentHeight: contentColumn.height
+        contentWidth: contentColumn.width + 1 // for trailing separator
```
Sorry, what do you mean? I don't really see how the code here relates to rounding. Did you mean a different line?

vpereverzev replied:
Yeah, ignore this one. I was reviewing this at deep night.

cbjeukendrup commented on src/framework/audio/qml/MuseScore/Audio/KnobControl.qml:
```diff
@@ -159,4 +159,52 @@ Dial {
+            root.newValueRequested(0)
+        }
+
+        // The MouseArea steals mouse press events from the slider.
```
I agree we can do this in a future PR. Maybe we should do this via C++, like the QML components also do. It would be great if we could subclass QML components, but we can't because they're in private headers...

vpereverzev replied:
If you're speaking about QtQuick.Controls, then there is no need to deal with this library here. The logic around this "new" component is pretty simple.

However, if it'll be necessary, we can wrap our own implementation into Control qml type (which is a base (C++) component for all QtQuick.Controls)

jessjwilliamson commented:
Hi! @vpereverzev is this ready to be merged please? :)

vpereverzev commented:
@cbjeukendrup please, resolve the conflicts and I'll merge this one

cbjeukendrup commented:
Done!
